### ISSUE TITLE
ScheduleAt function returns error "amqp: connection closed"

### ISSUE THREAD

**Issue by uch71109** [Azure/azure-service-bus-go#198]
We are using `ScheduleAt` and `CancelScheduled` function to handle delayed messages, the behaviour is working at first, but for a while (maybe one hour or two hours later)  the `ScheduleAt` doesn't work and we get an error `amqp: connection closed` 

It seems that it should try to reconnect to the server even if the connection is disconnected.

Could anyone please help this problem? Thanks.

SDK Version: `v0.10.6`

Expected behaviour:
The delayed message can be scheduled by `ScheduleAt` function without error

Actual behaviour:
It's okay at first but fails after a few hours later 

Reproduce step:
1. Create Queue instance https://github.com/Azure/azure-service-bus-go/blob/040d6776eae5108f83235d70063863db91822e88/queue.go#L142
2. Send delayed message https://github.com/Azure/azure-service-bus-go/blob/040d6776eae5108f83235d70063863db91822e88/entity.go#L215
2. After a few hours later, use the same queue instance to send message, then get `amqp: connection closed` error

**Comment by vgtswork** [comment]
Similar behivor here. Initially (few weeks ago) the error was that ```amqp: connection closed``` but now after few hours (usually 2hrs) i get a different error:

```unhandled error link kkkkkkkk-xxxx-yyyy-zzzz-ae293207a893: status code 401 and description: Unauthorized access. 'Send' claim(s) are required to perform this operation. Resource: 'sb://mytesttenant-westeurope-sevicebus-namespace.servicebus.windows.net/webhook-send/$management'. TrackingId:kkkkkkkk-xxxx-tttt-zzzz-38a45170fb24_G7, SystemTracker:NoSystemTracker, Timestamp:2020-12-07T19:31:07```

Obviusly 'Send' authorization is set for that shared policy key, and restarting the service fixes the error - then after 2hrs here we go again.

SDK Version ```v0.10.6```

At first I thought the cause might be some sort of "timeout" (in dev env we process very few messages) - so i tried to implement a simple "heartbeat" logic, a goroutine that peeks a message every 5 minutes to keep the connection alive. The result was actually the same, but this time the 401 error complains about 'Listen' claim(s) and not 'Send'.

**Important thing to note**:
* I uses the same Queue instance as listener and sender. The listener mostly never drops, NO problem with that, i continue to receive messages on the queue. The problem is only in the ```ScheduleAt``` phase
* Seems like this problem is a recent one - i never got that errore before, first time was ~1 month ago.

Any hints or help with that?

**Comment by devigned** [comment]
Sounds like the token used to establish the connection may have expired, and was not refreshed. 

When a SAS token is created, it is given a 2 hour expiry: https://github.com/Azure/azure-amqp-common-go/blob/6b7304f631077f049a7249a2f36fe87e5c27b674/sas/sas.go#L118

Might be a bug on connecting auth refreshing.

**Comment by vgtswork** [comment]
@devigned i'm doing some testing in local and seems like this is not the problem. Obviusly, i just started to dig into the package so, for sure, i'm missing something, but my first try was to lower in my local ```azure-amqp-common-go/sas/sas.go``` token lifespan to 10 seconds, and everything seems to be fine.

Every 5 mins ```periodicallyRefreshAuth``` is called and ```sas.GetToken``` seems to work fine.

Right now what i did is to enable the Jaeger opentracing in servicebus package and wait 2hrs both in my local and remote (dev) env - as soon as the error will pop again let's see if i can see some more error/tracing of that.

The huge problem is that bug is the repeatability, i need to wait those hrs before get the error again :(

If you have any other hint would be very nice, the package seems to work fine except from this ScheduleAt, that in my case is a very important piece of the puzzle.

**Comment by devigned** [comment]
> ScheduleAt, that in my case is a very important piece of the puzzle

Definitely a key differentiated feature of Service Bus. I'm sure determining a root cause and fix will be impactful to many.

**Comment by devigned** [comment]
@jhendrixMSFT or @catalinaperalta any chance you could ping a contact in Service Bus and see if they can comment on the issue? This might be something happening on the service side.

**Comment by vgtswork** [comment]
I made some small progress - im not sure about the 401 claim error but regarding the ```amqp: connection closed``` i made some tests.

I really don't know the root cause, but i think that after some time the ```rpcClient.client``` (```amqp.Client```) connection dies and the instance is not set to nil - this leads to some problems.

So, briefly, this happens:
* ```sendingEntity.ScheduleAt``` is called and calls ```se.entity.GetRPCClient```
https://github.com/Azure/azure-service-bus-go/blob/705d23958eb9c000582c03ba30413dc7e35eb25d/entity.go#L219
* then ```e.ensureRPCClient``` (entity) is called - that function just checks if the rpc client is not nil, otherwise create a new client with ```newRPCClient```. The problem is exactly there, if the connection dies or is closed the object is not set to nil and so everything goes ahead.
https://github.com/Azure/azure-service-bus-go/blob/705d23958eb9c000582c03ba30413dc7e35eb25d/entity.go#L250
* ```client.ScheduleAt``` (rpc client) is called
https://github.com/Azure/azure-service-bus-go/blob/705d23958eb9c000582c03ba30413dc7e35eb25d/entity.go#L225
*  ```ensureConn``` method should then ensure the connection
https://github.com/Azure/azure-service-bus-go/blob/705d23958eb9c000582c03ba30413dc7e35eb25d/rpc.go#L468
but there we go with the second point of failure where the problem can happen
https://github.com/Azure/azure-service-bus-go/blob/705d23958eb9c000582c03ba30413dc7e35eb25d/rpc.go#L88
there the method only checks if r.client is not nil. If the connection is dead or closed ```ensureConn``` just returns nil error and let the code continue.
* ```rpcClient.ScheduleAt``` now can fail in both
https://github.com/Azure/azure-service-bus-go/blob/705d23958eb9c000582c03ba30413dc7e35eb25d/rpc.go#L531
and
https://github.com/Azure/azure-service-bus-go/blob/705d23958eb9c000582c03ba30413dc7e35eb25d/rpc.go#L536
but will probably fail first in ```NewLink``` returning ```amqp: connection closed```

To replicate: just call ```rpcClient.Close()``` (or ```r.client.Close()```, depends on the function you are in) in a goroutine or somewhere in the middle of the code ScheduleAt code (i tried before the ```NewLink```). You will then get this error in the first ScheduleAt and all the following

TLDR/NOTE:
* "ensure" methods (both ```ensureRPCClient``` and ```ensureConn```) just checks for the client not to be nil, i think we should extend with some real connection checks there and recover if the connection is dead.
* at the moment i have 0 ideas about the root cause of the disconnection, i tried with lowering the token lifespan (even to 1s) but nothing happens
* i wanted to implement an rpc (amqp.Client in github.com/Azure/go-amqp package) connection check, but only exposed methods are ```Close``` and ```NewSession```
https://github.com/Azure/go-amqp/blob/6a20a079aa598260ec97857517635d5b02a20b6a/client.go#L31
but i probably need to dig better.

Im not sure if what i found is correct, but is something

**Comment by vgtswork** [comment]
small note on the above commet: update the SDK version to ```v0.10.7```

**Comment by devigned** [comment]
@vgtswork, if I recall, there wasn't a good way to check if the connection had been closed. I think with a bit of work at the amqp client level that could be built. 

As a stop gap, perhaps, try rebuilding the client if you run into an error here:
https://github.com/Azure/azure-service-bus-go/blob/705d23958eb9c000582c03ba30413dc7e35eb25d/entity.go#L225

The queue and topic sender will rebuild the client if they encounter an error.
https://github.com/Azure/azure-service-bus-go/blob/705d23958eb9c000582c03ba30413dc7e35eb25d/sender.go#L235-L266

**Comment by vgtswork** [comment]
@devigned good point, that can explains why the classic "sender" works and the ScheduleAt fails :)

Hope i will have a chance to step back on this task today and do some testing with your suggestions, otherwise i will do some testing tomorrow and update you guys there

Thanks a lot!

**Comment by SpaWn2KiLl** [comment]
I'm facing the same issue of receiving 401 - Unauthorized when calling ScheduleAt function after a few hours of my container being deployed.

Are you planning on changing the error handling for ScheduleAt just like Send does? It would be very helpful and so I don't need to put that logic into my application.

**Comment by vgtswork** [comment]
@SpaWn2KiLl hope i will be able to push a quick workaround PR soon, unfortunately had no time during this month :( i made some quick tests in local using the rebuild logic and everything seems to work.

Hope i can step back to this task next week - meanwhile if maintainers have some official/better solution would be great! :)

I think it would be great to find the root cause of the problem, it could be something caused by token expiration, but after some tests i failed to detect where it happens unfortunately.

**Comment by jhendrixMSFT** [comment]
@vgtswork do you have a simple repro program you could share?

**Comment by SpaWn2KiLl** [comment]
@vgtswork any news? It would be nice if you could share your progress :)

**Comment by vgtswork** [comment]
@SpaWn2KiLl @jhendrixMSFT im so sorry to both of you guys for my delay on news/answers :( 

unfortunately im right now very busy in some other tasks at work and had no chance to get back on this one. Hopefully i should very soon be able to step back and share what I've done. Consider that i had no time myself to deploy my fix proposal to my dev env, i just worked in local and made some quick assumpion.

Will ping you as soon as i have a forked repo with my quickfix proposal.

Thanks a lot!

**Comment by jhendrixMSFT** [comment]
I've been digging into this with @catalinaperalta, one thing that stands out to us is that SAS token refresh logic swallows any errors during refresh (they are logged but that's it).  Our best guess at this point is the refresh fails for some reason so the token is never updated on the service side and ends up expiring.

If anybody has tracing logs to confirm/deny this hypothesis that would be most helpful.  We're also trying to fab up a repro for this.

**Comment by jhendrixMSFT** [comment]
Also, due to clock skew and other "weirdness" of service bus, you can't have a SAS token with a lifespan of less than ~10 minutes.

**Comment by vgtswork** [comment]
@jhendrixMSFT i had a chance to try in my dev env the workaround i was working on but unfortunately it's not working... i tried to rebuild the connection in ScheduleAt in a similar way Sender (sender.go https://github.com/Azure/azure-service-bus-go/blob/705d23958eb9c000582c03ba30413dc7e35eb25d/sender.go#L235-L266) is doing.
I was initially sure it was working but doing some better long run testing after few hours it seems to die with the ```Unauthorized access. 'Sender' claim(s) are required```. 

So yeah, i suppose he best way to fix this is to focus on https://github.com/Azure/azure-service-bus-go/issues/155 issue

**Comment by jhendrixMSFT** [comment]
@vgtswork thanks for the update and the data point.  We'll keep digging.

**Comment by jhendrixMSFT** [comment]
One thing I just realized is that `ScheduleAt()` doesn't call `periodicallyRefreshAuth()`, so if you only call `ScheduleAt()` on the queue the credentials will expire after two hours.

**Comment by vgtswork** [comment]
@jhendrixMSFT in my service i actually have just a listener and and reschedule messages using ```ScheduleAt()```, nothing more. 

Am i wrong or the listener also calls ```periodicallyRefreshAuth()```?

**Comment by jhendrixMSFT** [comment]
`ScheduleAt()` for sure doesn't call this.  There is a `rpcClient.periodicallyRefreshAuth()` but it's not actually called (I will fix this).

Are you calling any other APIs?  If so they might enable periodic refresh (sample/repro code would help).

**Comment by jhendrixMSFT** [comment]
I have a change to consolidate the refresh logic which also ensures auto-refresh is enabled for all code-paths that require authentication.  I've tested it for the simple case and it's working but am testing further.  If anybody here can also test it out it would be much appreciated.

https://github.com/Azure/azure-service-bus-go/pull/205

**Comment by vgtswork** [comment]
@jhendrixMSFT in this particular service that fails after 2h i call only ```queue.NewReceiver(ctx)``` (```*servicebus.Queue``` type) and the ```Listen(ctx, handler)``` on that receiver.
Then inside the listener i only use message dispositions and ```ScheduleAt()``` to requeue a delayed message

**Comment by jhendrixMSFT** [comment]
So the call to `queue.NewReceiver()` does start auto-refresh for auth.  Presumably your code looks something like this (excuse the terse error-handling).

```go
q, err := ns.NewQueue("helloworld")
if err != nil {
	panic(err)
}
rec, err := q.NewReceiver(context.Background())
rec.Listen(context.Background(), servicebus.HandlerFunc(func(ctx context.Context, msg *servicebus.Message) error {
	// do something with msg.....
	if _, err := q.ScheduleAt(context.Background(), time.Now(), &servicebus.Message{}); err != nil {
		return err
	}
	return nil
}))
```

Going back to the beginning, I think there are two distinct issues here.

The first error, `amqp: connection closed`, when calling `q.ScheduleAt()` as reported by @uch71109.  I suspect this is due to no recovery mechanism on the `rpcClient` like there is with senders and receivers.  This one still needs a fix.

The second error, `authentication failed`, as reported by @vgtswork, which would appear to be an issue with the authentication refresh logic (in some cases it wasn't even enabled).  https://github.com/Azure/azure-service-bus-go/pull/205 fixes the case where refresh wasn't being enabled, however I still feel there's another piece to this puzzle that I don't have yet so I'm not totally convinced this fixes all auth cases.

**Comment by jhendrixMSFT** [comment]
I'm testing out a recovery mechanism for `rpcClient` in https://github.com/Azure/azure-service-bus-go/pull/206

**Comment by jhendrixMSFT** [comment]
This should now be fixed in `v0.10.10`.  Thanks everybody for the help.
